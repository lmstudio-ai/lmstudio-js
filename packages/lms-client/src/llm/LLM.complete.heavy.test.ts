import { type LLM, LMStudioClient } from "../index.js";
import { ensureHeavyTestsEnvironment, llmTestingModel } from "../shared.heavy.test.js";

describe("LLM", () => {
  let client: LMStudioClient;
  let model: LLM;
  beforeAll(async () => {
    client = new LMStudioClient();
    await ensureHeavyTestsEnvironment(client);
  });
  beforeEach(async () => {
    model = await client.llm.model(llmTestingModel, { verbose: false });
  });
  describe(".complete", () => {
    it("should work without streaming", async () => {
      const result = await model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 3,
        stopStrings: [";"],
      });
      expect(result.content).toMatchInlineSnapshot(`"4"`);
      expect(result.stats).toMatchSnapshot({
        numGpuLayers: expect.any(Number),
        timeToFirstTokenSec: expect.any(Number),
        tokensPerSecond: expect.any(Number),
      });
      expect(result.modelInfo).toMatchSnapshot({
        identifier: expect.any(String),
        instanceReference: expect.any(String),
        modelKey: expect.any(String),
      });
      expect(result.roundIndex).toEqual(0);
    });
    it("should work with streaming", async () => {
      const prediction = model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 3,
        stopStrings: [";"],
      });
      const fragments = [];
      for await (const fragment of prediction) {
        fragments.push(fragment);
      }
      expect(fragments).toMatchSnapshot();
      const result = await prediction.result();
      expect(result.content).toEqual("4");
      expect(result.stats).toMatchSnapshot({
        numGpuLayers: expect.any(Number),
        timeToFirstTokenSec: expect.any(Number),
        tokensPerSecond: expect.any(Number),
      });
      expect(result.modelInfo).toMatchSnapshot({
        identifier: expect.any(String),
        instanceReference: expect.any(String),
        modelKey: expect.any(String),
      });
    });
    it("should allow cancel via .cancel()", async () => {
      const prediction = model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 50,
      });
      prediction.cancel();
      const result = await prediction.result();
      expect(result.stats.stopReason).toEqual("userStopped");
    });
    it("should allow cancel via abort controller", async () => {
      const controller = new AbortController();
      const prediction = model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 50,
        signal: controller.signal,
      });
      controller.abort();
      const result = await prediction.result();
      expect(result.stats.stopReason).toEqual("userStopped");
    });
    it("should call onFirstToken", async () => {
      const onFirstToken = jest.fn();
      await model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 1,
        stopStrings: [";"],
        onFirstToken,
      });
      expect(onFirstToken).toHaveBeenCalled();
    });
    it("should call onPromptProcessingProgress", async () => {
      const onPromptProcessingProgress = jest.fn();
      await model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 1,
        stopStrings: [";"],
        onPromptProcessingProgress,
      });
      expect(onPromptProcessingProgress).toHaveBeenCalledWith(0);
      expect(onPromptProcessingProgress).toHaveBeenCalledWith(1);
    });
    it("should call onPredictionFragment", async () => {
      const onPredictionFragment = jest.fn();
      await model.complete("1 + 1 = 2; 2 + 2 = ", {
        temperature: 0,
        maxTokens: 3,
        onPredictionFragment,
      });
      const calls = onPredictionFragment.mock.calls;
      // Note: The this seem to contain wrong results as it only has 2 tokens. It might be an engine
      // but. We can correct the snapshot once the engine bug is fixed.
      expect(calls).toMatchSnapshot();
    });
  });
});
