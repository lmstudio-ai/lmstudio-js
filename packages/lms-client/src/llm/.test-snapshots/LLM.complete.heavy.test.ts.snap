// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LLM.complete should call onPredictionFragment 1`] = `
[
  [
    {
      "containsDrafted": false,
      "content": "4",
      "reasoningType": "none",
      "tokensCount": 1,
    },
  ],
  [
    {
      "containsDrafted": false,
      "content": ";",
      "reasoningType": "none",
      "tokensCount": 1,
    },
  ],
]
`;

exports[`LLM.complete should support structured prediction with JSON schema 1`] = `
{
  "answer": 2,
}
`;

exports[`LLM.complete should support structured prediction with zod schema 1`] = `
{
  "answer": 2,
}
`;

exports[`LLM.complete should support structured prediction with zod schema 2`] = `
{
  "answer": 2,
}
`;

exports[`LLM.complete should work with streaming 1`] = `
[
  {
    "containsDrafted": false,
    "content": "4",
    "reasoningType": "none",
    "tokensCount": 1,
  },
]
`;

exports[`LLM.complete should work with streaming 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 1,
  "promptTokensCount": 15,
  "stopReason": "stopStringFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 16,
}
`;

exports[`LLM.complete should work with streaming 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "0.5B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM.complete should work without streaming 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 1,
  "promptTokensCount": 15,
  "stopReason": "stopStringFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 16,
}
`;

exports[`LLM.complete should work without streaming 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "0.5B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;
